### PR TITLE
[Arith] linear system and equation solver

### PR DESCRIPTION
**Pull request by yzhliu**
Co-authored-by: Sergei Grechanik <sergei.grechanik+h@gmail.com>

Linear system data structures and linear equations solver (via Smith Normal Form)

this is part of [tensor level autodiff](https://discuss.tvm.ai/t/rfc-bring-in-tensor-expression-autodiff). will be used by https://github.com/apache/incubator-tvm/pull/5121

Originally implemented by @sgrechanik-h in https://github.com/apache/incubator-tvm/pull/2634 I modified a bit to make it more mathematical. @sgrechanik-h please help to check if the math is correct and whether `SuperSimplify` is required (I use `tir::Simplify` instead in this PR, I think it is only required in condition lift)

@sgrechanik-h @tqchen @MarisaKirisame @junrushao1994 @hzfan Please review

### CODE REVIEW

**Comment by tqchen** [comment]
NOTE: please try to rewrite to directly use the auxiliary class arith::Analyzer instead of the Simplify function, so that the temp data structure can be created  only once

**Comment by yzhliu** [comment]
@tqchen changed to use arith::Analyzer

File: include/tvm/arith/linear_system.h [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+namespace tvm {
+namespace arith {
+
+using tir::Var;
+using tir::VarNode;
+using tir::IterVar;
+
+/*!
+ * \brief Represent a linear system including variables, their ranges and
+ *        the linear relations between them (either equations or inequalities)
```
Comment (sergei-grechanik) [inline include/tvm/arith/linear_system.h:41]: I think it's worth mentioning that the system is integer.

Also I'm not sure if LinearSystem is a good name, because it may be useful to have non-linear (in)equalities in `relations` (which doesn't break the algorithms that work on the linear part).

File: include/tvm/arith/linear_system.h [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                                Map<Var, PrimExpr> dst_to_src);
+
+  TVM_DEFINE_OBJECT_REF_METHODS(LinearSystemTransform, ObjectRef, LinearSystemTransformNode);
+};
+
+/*!
+ * \brief Obtain Smith Normal Form of linear equation A x = y.
+ *        Smith Normal Form of matrix A_{mxn} is S_{mxn} = U_{mxm} A_{mxn} V_{nxn},
+ *        in which S_{mxn} is diag(s1, s2, ..., sr, 0, ..., 0),
+ *        such that si | s_{i+1} and r is the rank of A.
```
Comment (sergei-grechanik) [inline include/tvm/arith/linear_system.h:143]: I'm not sure, but I think the condition `si | s_{i+1}` is not guaranteed by this implementation, so it's not fully fledged Smith Normal Form (which shouldn't be a problem for our end goal).

File: tests/python/unittest/test_arith_solve_linear_system.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    assert ir_pass.Equal(rel, False)
+    assert len(solution.src_to_dst) == 0
+    assert len(solution.dst_to_src) == 0
+
+
+if __name__ == "__main__":
+    test_unique_solution()
+    test_low_rank()
+    test_infer_range()
+    test_ill_formed()
```
Comment (sergei-grechanik) [inline tests/python/unittest/test_arith_solve_linear_system.py:91]: I would also recommend bringing in testing via random system generation, it helped me a lot in discovering subtle bugs.

**Comment by sergei-grechanik** [comment]
About `SuperSimplify`: it was a combination of the canonical simplifier and the rewriting simplifier that worked best for autodiff (I actually ran it with different combinations and rewrite->canonical->rewrite turned out to be the best, although canonical->rewrite was good too). I think currently the default Simplify function is rewrite->canonical, not sure if this order has a good justification.

File: tests/python/unittest/test_arith_solve_linear_system.py (hunk 3 above)
Comment (yzhliu) [reply tests/python/unittest/test_arith_solve_linear_system.py:91]: how did you generate random cases? random coefs or something smarter?

**Comment by yzhliu** [comment]
@sergei-grechanik do you have example that we have to end with rewrite?

File: src/arith/linear_system.cc [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+namespace arith {
+
+LinearSystem::LinearSystem(Array<Var> variables,
+                           Map<Var, Range> ranges,
+                           Array<PrimExpr> relations) {
+  ObjectPtr<LinearSystemNode> node = make_object<LinearSystemNode>();
+  node->variables = std::move(variables);
+  node->ranges = std::move(ranges);
+  node->relations = std::move(relations);
+  data_ = std::move(node);
```
Comment (MarisaKirisame) [inline src/arith/linear_system.cc:43]: what is data_?

File: src/arith/solve_linear_equation.cc [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+}
+
+TVM_REGISTER_GLOBAL("arith.SolveEquations")
+.set_body([](TVMArgs args, TVMRetValue *ret) {
+    if (args.size() == 1) {
+      *ret = SolveEquations(args[0]);
+    } else if (args.size() == 3) {
+      LinearSystem problem(args[0], args[1], args[2]);
+      *ret = SolveEquations(problem);
+    }
```
Comment (MarisaKirisame) [inline src/arith/solve_linear_equation.cc:472]: should fail explicitly for args of other size

**Comment by tqchen** [comment]
cc @Hzfengsy @spectrometerHBH @merrymercy please also take a look if you have time, should be relevant to your areas

**Comment by sergei-grechanik** [comment]
@yzhliu Unfortunately I don't have an example. It's written in my notes that "we should end with rw because it factors multipliers out". If it's only that, it's probably not that important.

File: tests/python/unittest/test_arith_solve_linear_system.py (hunk 3 above)
Comment (sergei-grechanik) [reply tests/python/unittest/test_arith_solve_linear_system.py:91]: Yes, it's exactly random coefs. I had to tune coefficient ranges manually because sometimes it resulted in integer overflow.
https://github.com/sergei-grechanik/tvm/blob/8cad2d1e62272b3e192bfe08b896e07bc9550e94/tests/python/unittest/test_pass_zero_elimination.py#L301

File: src/arith/solve_linear_equation.cc [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+          analyzer_problem.Simplify(eq->a - eq->b),
+          system_to_solve->variables);
+      if (!coeffs.empty()) {
+        std::vector<int64_t> row;
+        for (size_t j = 0; j < coeffs.size() - 1; ++j) {
+          PrimExpr c = coeffs[j];
+          if (const IntImmNode* ic = c.as<IntImmNode>()) {
+            row.push_back(ic->value);
+          } else {
+            // elements in matrix S V must be integers
```
Comment (hzfan) [inline src/arith/solve_linear_equation.cc:313]: Seems that an equation is ignored, shall we emit a warning here?

File: src/arith/solve_linear_equation.cc [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+  // initialize V^{-1} x as x
+  for (size_t i = 0; i < num_vars; ++i) {
+    V.emplace_back(num_vars);
+    V.back()[i] = 1;
+    V_inv_x.push_back(system_to_solve->variables[i]);
+  }
+
+  // Transform formulas into rows of the matrix
+  // S_{mxn} V^{-1}_{nxn} x_{nx1} = U y, in which n is # of variables
+  // here we initialize S_{mxn} to be A, U to be identity matrix.
```
Comment (hzfan) [inline src/arith/solve_linear_equation.cc:299]: V to be identity matrix?

File: src/arith/solve_linear_equation.cc [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+    }
+  }
+
+  Array<PrimExpr> solution_for_V_inv_x;
+  // Now create new variables or directly solve the equations
+  // suppose the rank of A is r, aka r = # of non-zeros in S
+  // the solution of S_{mxn} V^{-1}_{nxn} x_{nx1} = U b
+  // is
+  // x = (pseudo-inverse of A) b + K_{n, n-r} z_{n-r}
+  //   = V_{n,n} S^{-1}_{n,m} (Ub)_{mxn} + K_{n, n-r} z_{n-r}
```
Comment (hzfan) [inline src/arith/solve_linear_equation.cc:383]: What about `_{nxm}` and `Uy` to be consistent?

File: include/tvm/arith/linear_system.h (hunk 1 above)
Comment (yzhliu) [reply include/tvm/arith/linear_system.h:41]: good suggestion. I have several proposals, how about Constraints, RelationSet, IntRelationSet, or InequalitySet (if we make term "inequality" to also include equations)?

Comment (sergei-grechanik) [reply include/tvm/arith/linear_system.h:41]: IntegerSet or IntegerSystem maybe?

File: src/arith/linear_system.cc (hunk 4 above)
Comment (yzhliu) [reply src/arith/linear_system.cc:43]: it's ObjectRef->data_ whose type is ObjectPtr<Object>

File: include/tvm/arith/linear_system.h (hunk 1 above)
Comment (yzhliu) [reply include/tvm/arith/linear_system.h:41]: sounds good. There's IntSet in tvm, so I'll do IntegerSystem.

File: src/arith/solve_linear_equation.cc (hunk 6 above)
Comment (yzhliu) [reply src/arith/solve_linear_equation.cc:313]: good suggestion.

File: src/arith/solve_linear_equation.cc (hunk 7 above)
Comment (yzhliu) [reply src/arith/solve_linear_equation.cc:299]: V was described above. here what I mean is
U^{-1} S V^{-1} x = y = [coefs] => S V^{-1} x = U [coefs]
which implies U is identity.

File: include/tvm/arith/linear_system.h (hunk 1 above)
Comment (yzhliu) [reply include/tvm/arith/linear_system.h:41]: think it a bit more, should we use a more general name? as in solver, we will check the relation and ignore the non-integer equations, which implies it is valid to contain non-integer relations inside. @sergei-grechanik

Comment (sergei-grechanik) [reply include/tvm/arith/linear_system.h]: If variables are integer, I think it's still ok to call it IntegerSystem. But if we allow non-integer variables (I doubt it is a good idea), then I don't know, Set and System sound too unspecific, Constraints may be ok.

**Comment by yzhliu** [comment]
@sergei-grechanik I rename it to IntConstraints, and also add checker to ensure the variables to be integers. please take a review again.

File: tests/python/unittest/test_arith_solve_linear_system.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import random
+import numpy as np
+import tvm
+from tvm import te, arith, ir, tir
+
+
+def run_expr(expr, vranges):
```
Comment (sergei-grechanik) [inline tests/python/unittest/test_arith_solve_linear_system.py:23]: Some comment like "Evaluate expr for every value of free variables given by vranges and return the tensor of results.".
Also I think these functions are quite useful and need to be moved somewhere like tvm.testing (but probably not in this PR, so may be add a TODO).

File: tests/python/unittest/test_arith_solve_linear_system.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+
+    A = te.compute([r.extent.value for v, r in vranges.items()], _compute_body)
+    args = [tvm.nd.empty(A.shape, A.dtype)]
+    sch = te.create_schedule(A.op)
+    mod = tvm.build(sch, [A])
+    mod(*args)
+    return args[0].asnumpy()
+
+
+def check_bruteforce(bool_expr, vranges, cond=None):
```
Comment (sergei-grechanik) [inline tests/python/unittest/test_arith_solve_linear_system.py:36]: Some comment like "Check that bool_expr holds given the condition cond for every value of free variables from vranges".

File: src/arith/solve_linear_equation.cc [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+    new_relation = analyzer_problem.Simplify(new_relation);
+    if (tir::is_const_int(new_relation, 0)) {
+      // unable to solve the system.
+      return IntConstraintsTransform(
+          system_to_solve,
+          IntConstraints(
+              /*variables=*/{},
+              /*ranges=*/{},
+              /*relations=*/{te::make_zero(DataType::Bool())}),
+          {}, {});
```
Comment (sergei-grechanik) [inline src/arith/solve_linear_equation.cc:373]: I'm not sure, but I think src_to_dst still should map original variables to some dummy values (zero probably), this should prevent edge cases somewhere.

File: tests/python/unittest/test_arith_solve_linear_system.py [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+        counterex = [(str(v), i + r.min) for (v, r), i in zip(vranges.items(), indices)]
+        counterex = sorted(counterex, key=lambda x: x[0])
+        counterex = ", ".join([v + " = " + str(i) for v, i in counterex])
+        raise AssertionError("Expression {}\nis not true on {}\n"
+                             "Counterexample: {}"
+                             .format(tir.ir_pass.CanonicalSimplify(bool_expr), vranges, counterex))
+
+
+def check_solution(solution, vranges={}):
+    def _check_forward(formula1, formula2, varmap, backvarmap):
```
Comment (sergei-grechanik) [inline tests/python/unittest/test_arith_solve_linear_system.py:52]: formula1 -> constraints1 or something like that

File: tests/python/unittest/test_arith_solve_linear_system.py [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+        indices = list(np.argwhere(res == 0)[0])
+        counterex = [(str(v), i + r.min) for (v, r), i in zip(vranges.items(), indices)]
+        counterex = sorted(counterex, key=lambda x: x[0])
+        counterex = ", ".join([v + " = " + str(i) for v, i in counterex])
+        raise AssertionError("Expression {}\nis not true on {}\n"
+                             "Counterexample: {}"
+                             .format(tir.ir_pass.CanonicalSimplify(bool_expr), vranges, counterex))
+
+
+def check_solution(solution, vranges={}):
```
Comment (sergei-grechanik) [inline tests/python/unittest/test_arith_solve_linear_system.py:51]: Something like "Check that solution is a bijective transformation"

File: tests/python/unittest/test_arith_solve_linear_system.py [hunk 14]
```diff
@@ -14,9 +14,130 @@
...
+        # not solvable, skip
+        return
+    _check_forward(solution.src, solution.dst,
+                   solution.src_to_dst, solution.dst_to_src)
+    _check_forward(solution.dst, solution.src,
+                   solution.dst_to_src, solution.src_to_dst)
+
+
+def test_solution_consistency():
+    random.seed(0)
```
[the remark this answers is not in the document]
Comment (yzhliu) [reply tests/python/unittest/test_arith_solve_linear_system.py:87]: how about we always use random seed, and print out the seed so that we will be able to reproduce once triggered?

[the remark this answers is not in the document]
Comment (sergei-grechanik) [reply tests/python/unittest/test_arith_solve_linear_system.py:87]: Personally, I think it is a good idea. I'm just worried that people won't know what to do with such failure and will just complain about flaky tests instead of reporting the unlucky seed. Maybe print some message like "this test is intentionally nondeterministic, if it fails please report it together with this seed"?

File: include/tvm/arith/int_solver.h [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+ *        NOTE: Although in standard Smith Normal Form the diagonal elements satisfy
+ *              s_i | s_{i+1} (| means divides), the implement here does not guarantee it.
+ *        U_{mxm} and V_{nxn} are invertible matrices.
+ *        This function modifies \p S to be S_{mxn}, \p V to be V_{nxn},
+ *        \p y to be U_{mxm} y_{mx1} and \p x to be V^{-1} x.
+ * \param S  the original A_{mxn}, it will be modified to S_{mxn}
+ * \param V  an identity matrix, it will be modified to V_{nxn}
+ * \param x  the x in A x = y. it will be modified to V^{-1}_{nxn} x_{nx1}
+ * \param y  the y in A x = y. it will be modified to U_{mxm} y_{mx1}
+ */
```
Comment (sergei-grechanik) [inline include/tvm/arith/int_solver.h:153]: I was thinking about this and came to conclusion that computing the proper Smith normal form may improve stability of automatic differentiation (I mean generating the same gradient code for slightly different but equivalent input code). However, I'm not sure if it's worth it (and not in this PR of course).

Comment (yzhliu) [reply include/tvm/arith/int_solver.h:153]: I see. I'll put a todo here.

File: src/arith/solve_linear_equation.cc (hunk 11 above)
Comment (yzhliu) [reply src/arith/solve_linear_equation.cc:373]: In `check_solution` test I did see such dummy mapping helps to pass the check. while I'm not sure whether this is the right thing to do, since `src_to_dst` is normally defined to provide the "solution". How about we leave it empty here and revisit if later we find it is necessary?
For now in test I do this: https://github.com/apache/incubator-tvm/pull/5171/files#diff-b8ec9195578ec265834eddba2d5f5225R77-R79

File: tests/python/unittest/test_arith_solve_linear_system.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (yzhliu) [reply tests/python/unittest/test_arith_solve_linear_system.py]: good idea.

File: src/arith/solve_linear_equation.cc (hunk 11 above)
Comment (sergei-grechanik) [reply src/arith/solve_linear_equation.cc:373]: Ok

File: tests/python/unittest/test_arith_solve_linear_system.py [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+        return
+    _check_forward(solution.src, solution.dst,
+                   solution.src_to_dst, solution.dst_to_src)
+    _check_forward(solution.dst, solution.src,
+                   solution.dst_to_src, solution.src_to_dst)
+
+
+def test_solution_consistency(capsys):
+    seed = random.randrange(sys.maxsize)
+    with capsys.disabled():
```
Comment (sergei-grechanik) [inline tests/python/unittest/test_arith_solve_linear_system.py:99]: Not sure, may be it would be better to leave it captured so that the message is shown on failure?

File: python/tvm/arith/int_solver.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    dst_to_src : Map[tvm.tir.Var, tvm.ir.PrimExpr]
+        mapping from variables in the dst to the variables in the src,
+        e.g., {m -> a, n -> -b}
+    """
+    def __init__(self, src, dst, src_to_dst, dst_to_src):
+        self.__init_handle_by_constructor__(
+            _ffi_api.IntConstraintsTransform, src, dst, src_to_dst, dst_to_src)
+
+
+def solve_linear_equations(equations, variables, ranges):
```
Comment (sergei-grechanik) [inline python/tvm/arith/int_solver.py:74]: defaults: `variables=None, ranges=None` (and mark them as optional in the docstring)

And may be add a test for this mode of calling

File: include/tvm/arith/int_solver.h [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+/*!
+ * \brief Represent integer constrains including (integer) variables, their ranges and
+ *        the relations between them (either equations or inequalities).
+ * \sa LinearSystem
+ */
+class IntConstraintsNode : public Object {
+ public:
+  // e.g., \alpha, \beta, must be integers
+  Array<Var> variables;
+  // e.g., 1 <= \alpha <= N, etc.
```
Comment (sergei-grechanik) [inline include/tvm/arith/int_solver.h:48]: Add a comment that it is absolutely ok to include ranges for parameters (variables that are not in `variables`) in this map.

File: python/tvm/arith/int_solver.py (hunk 17 above)
Comment (yzhliu) [reply python/tvm/arith/int_solver.py]: done

File: include/tvm/arith/int_solver.h (hunk 18 above)
Comment (yzhliu) [reply include/tvm/arith/int_solver.h:48]: added

File: tests/python/unittest/test_arith_solve_linear_system.py (hunk 16 above)
Comment (yzhliu) [reply tests/python/unittest/test_arith_solve_linear_system.py]: yup, removed.

**Comment by tqchen** [comment]
Thanks @yzhliu @sgrechanik-h @hzfan @MarisaKirisame
